core/src/validator.rs

```diff
@@ -1926,12 +1958,24 @@ mod tests {
 
         // Each validator can exit in parallel to speed many sequential calls to join`
         validators.iter_mut().for_each(|v| v.exit());
-        // While join is called sequentially, the above exit call notified all the
-        // validators to exit from all their threads
-        validators.into_iter().for_each(|validator| {
-            validator.join();
+
+        // spawn a new thread to wait for the join of the validator
+        let (sender, receiver) = unbounded();
+        let _ = thread::spawn(move || {
+            validators.into_iter().for_each(|validator| {
+                validator.join();
+            });
+            sender.send(()).unwrap();
         });
 
+        // timeout of 30s for shutting down the validators
+        if receiver.recv_timeout(Duration::from_secs(30)).is_err() {
+            panic!(
+                "timeout for shutting down validators: {:?}",
+                Backtrace::new()
```

Comment:
what does this add over the backtrace we'll get from panicking anyway?

Seems like best ROI would be to get the backtrace from every _other_ running thread, but I don't see a straight forward way to do that

Reply:
yeah. removed.